src/plugins/dashboard/public/application/utils/use/use_dashboard_container.tsx

```diff
@@ -230,21 +299,82 @@ const createDashboardEmbeddable = async (
             ) {
               // Add filters modifies the object passed to it, hence the clone deep.
-              filterManager.addFilters(_.cloneDeep(container.getInput().filters));
+              filterManager.addFilters(cloneDeep(container.getInput().filters));
 
+              // TODO: investigate if this is needed
               /* dashboardStateManager.applyFilters(
                   $scope.model.query,
                   container.getInput().filters
-                );*/
+              );*/
 
               appState.transitions.set('query', queryStringManager.getQuery());
             }
-            // TODO: triggered when dashboard embeddable container has changes, and update the appState
-            // handleDashboardContainerChanges(container, appState, dashboardServices);
+            // triggered when dashboard embeddable container has changes, and update the appState
+            handleDashboardContainerChanges(container, appState, dashboardServices);
           });
           return dashboardContainer;
         }
       });
   }
   return undefined;
 };
+
+const handleDashboardContainerChanges = (
+  dashboardContainer: DashboardContainer,
+  appState: DashboardAppStateContainer,
+  dashboardServices: DashboardServices
+) => {
+  let dirty = false;
+  let dirtyBecauseOfInitialStateMigration = false;
+  const appStateData = appState.getState();
+  const savedDashboardPanelMap: { [key: string]: SavedDashboardPanel } = {};
+  const { opensearchDashboardsVersion } = dashboardServices;
+  const input = dashboardContainer.getInput();
+  appStateData.panels.forEach((savedDashboardPanel) => {
+    if (input.panels[savedDashboardPanel.panelIndex] !== undefined) {
+      savedDashboardPanelMap[savedDashboardPanel.panelIndex] = savedDashboardPanel;
+    } else {
+      // A panel was deleted.
+      dirty = true;
+    }
+  });
+  const convertedPanelStateMap: { [key: string]: SavedDashboardPanel } = {};
+  Object.values(input.panels).forEach((panelState) => {
+    if (savedDashboardPanelMap[panelState.explicitInput.id] === undefined) {
+      dirty = true;
+    }
+    convertedPanelStateMap[panelState.explicitInput.id] = convertPanelStateToSavedDashboardPanel(
+      panelState,
+      opensearchDashboardsVersion
+    );
+    if (
+      !isEqual(
+        convertedPanelStateMap[panelState.explicitInput.id],
+        savedDashboardPanelMap[panelState.explicitInput.id]
+      )
+    ) {
+      // A panel was changed
+      dirty = true;
+      const oldVersion = savedDashboardPanelMap[panelState.explicitInput.id]?.version;
+      const newVersion = convertedPanelStateMap[panelState.explicitInput.id]?.version;
+      if (oldVersion && newVersion && oldVersion !== newVersion) {
+        dirtyBecauseOfInitialStateMigration = true;
+      }
+    }
+  });
+  if (dirty) {
+    appState.transitions.set('panels', Object.values(convertedPanelStateMap));
+    if (dirtyBecauseOfInitialStateMigration) {
+      // this.saveState({ replace: true });
```

Comment:
What is the logic here? why we need to check the situation where the version of a panel state in the dashboard container is different from the one in the app state?

Reply:
This is part of the old logics from the previous dashboard plugin. The logic here is if we detect a panel is removed, or changed or added, then dirty will be set to true. Then we will update the appState panels with the updated list of panels.